XMLDSIG from .NET Core.

vladimir-kazakov opened the pull request:
This PR brings last pieces of `System.Security.Cryptography.Xml` from .NET Core.

2 files couldn't be taken from .NET Core, though:

1. KeyInfoX509Data.cs, because the .NET Core version depends on the `System.Numerics` assembly (it uses the `BigInteger` type) and Mono can't use it. I noticed that .NET Framework does it differently - it uses a lightweight, internal version of a big integer (`BigInt`), which is available in /referencesource. So, I decided to take it from there. The rest of the code is just a copy-paste from the .NET Core version, using Mono's coding style.
1. SignedXml.cs, because in .NET Core its APIs are locked to .NET Framework 4.6.1. New APIs that were added after that are not available. Since this is not okay for Mono, I decided to copy-paste this file as well, including changes that happened after .NET Framework 4.6.1.

It's also worth mentioning that in .NET Framework 4.7.1, SHA-1 was deprecated and some default values were changed to [SHA-2 (256)](https://github.com/Microsoft/dotnet/blob/master/Documentation/compatibility/Change-SignedXML-and-SignedCMS-default-algorithms-to-SHA256.md). .NET Framework uses so called app context switches in order to decide whether the old or the new behavior should be used. .NET Core doesn't have such switches, and, if I understood everything correctly, Mono doesn't use them as well, always defaulting to the new behavior. That's what I did in this PR as well, but still, it would be interesting to know whether Mono is interested in supporting these app context switches. Perhaps somebody from Mono's maintainers could clarify this point.

Speaking of tests, after applying changes in this PR, a few of the tests started to fail. It turned out that this is because the old implementation was either incorrect or different compared to .NET Framework and .NET Core. The biggest problem was related to the fact that Mono was supporting truncated HMACs in signatures. Although it was according to [the XML Signature standard](https://www.w3.org/TR/xmldsig-core1/#sec-SignatureMethod), .NET Framework simply [rejects](https://referencesource.microsoft.com/#System.Security/system/security/cryptography/xml/signedxml.cs,576) such signatures, which is not against the standard. More info about this security issue can be found in [this article](https://www.w3.org/blog/2009/07/hmac-truncation-in-xml-signatu/).

marek-safar commented:
Add API_BIN_REFS := System.Numerics to the Makefile (after you bump api-snapshot submodule) and you should be able to use it.

I don't understand SignedXml.cs change, netcore is not locked to 4.6.1 what is missing?

vladimir-kazakov commented:
> Add API_BIN_REFS := System.Numerics to the Makefile (after you bump api-snapshot submodule) and you should be able to use it.

I get an error saying that `System.Numerics.dll` could not be found. I also don't see anything in the Makefile of `api-snapshot` related to this assembly. I'd add it myself, but even without any modifications, when I run `make`, I get a few errors like `error CS0619: 'Span<T>' is obsolete: 'Types with embedded references are not supported in this version of your compiler.'`

> I don't understand SignedXml.cs change, netcore is not locked to 4.6.1 what is missing?

A few public (in .NET Framework) constants [were made internal](https://github.com/dotnet/corefx/pull/19189#pullrequestreview-35663397) in .NET Core. In the end of this page there's an explanation why it was done.

marek-safar commented:
@vladimir-kazakov please try again I think I resolved the System.Numerics issue

marek-safar commented on mcs/class/referencesource/System/security/system/security/cryptography/x509/x509utils.cs:
```diff
@@ -23,6 +23,56 @@ namespace System.Security.Cryptography.X509Certificates {
 
     internal class X509Utils {
         private X509Utils () {}
+
```
Please don't rearrange referencesource code, it's hard to sync

vladimir-kazakov commented:
@marek-safar, now referencing System.Numerics in API_BIN_REFS works as expected and I'm able to use KeyInfoX509Data.cs from .NET Core.

But one of the tests started to fail because of the issue https://github.com/mono/mono/issues/6264. It seems that Mono previously was using only instances of [X509Certificate](https://github.com/mono/mono/blob/master/mcs/class/System.Security/System.Security.Cryptography.Xml/KeyInfoX509Data.cs#L54) and Mono's [own implementation](https://github.com/mono/mono/blob/master/mcs/class/System.Security/System.Security.Cryptography.Xml/KeyInfoX509Data.cs#L94) of it. At the same time, both .NET Framework and .NET Core are working with [X509Certificate2](https://github.com/dotnet/corefx/blob/master/src/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/KeyInfoX509Data.cs#L38) under the hood.

What would be the correct way to proceed? Should this PR wait until the mentioned issue will be fixed, or maybe the current approach could be used instead of waiting for the issue to be fixed? When I copy-pasted the implementation of KeyInfoX509Data, I left some code as it was in Mono before. It was doing the same things as needed, but a bit differently. I didn't notice any problems in this approach.

marek-safar commented:
commit apidiff

marek-safar commented:
Where do you see the failure?

vladimir-kazakov commented:
I see it locally, when I start using KeyInfoX509Data.cs from .NET Core. The [Constructor_X509CertificateByteArray_Null](https://github.com/mono/mono/blob/master/mcs/class/System.Security/Test/System.Security.Cryptography.Xml/KeyInfoX509DataTest.cs#L91) test fails.

Currently, in this PR all tests pass, because I haven't committed any changes yet.

marek-safar commented:
commit apidiff

vladimir-kazakov commented:
I rebased it on the top of the master today. I thought that fixing #6264 will fix the failing test, but it didn't. The test was still failing with the same exception message, although the stacktrace was different, pointing to the fact that there's another issue related to the usage / creation of empty X.509 certificates. I fixed this issue in a separate commit in this PR. With this change, all tests pass. Well, we can see failed checks, but it seems that it's something not related to my changes.

Speaking of the API diff, I somehow lost that commit, so there's a need to commit API changes again, in order to pass this check.

marek-safar commented:
commit apidiff

akoeplinger commented:
@monojenkins build Linux x64 mcs

akoeplinger commented:
@monojenkins build Linux x64 FullAOT

akoeplinger commented:
@monojenkins build Linux AArch64
